src/cluster/slot_migrate.cc

```diff
@@ -1130,3 +1148,164 @@ void SlotMigrator::resumeSyncCtx(const Status &migrate_result) {
+    if (catchUpIncrementalWAL()) {
+      break;
+    }
+    wal_incremental_seq = storage_->GetDB()->GetLatestSequenceNumber();
+    auto s = migrateIncrementalDataByRawKV(wal_incremental_seq, &batch_sender);
+    if (!s.IsOK()) {
+      return {Status::NotOK, fmt::format("migrate incremental data failed, {}", s.Msg())};
+    }
+    LOG(INFO) << fmt::format("[migrate] Migrated incremental data, epoch: {}, seq from {} to {}", epoch, wal_begin_seq_,
+                             wal_incremental_seq);
+    wal_begin_seq_ = wal_incremental_seq;
+    epoch++;
+  }
+
+  setForbiddenSlot(migrating_slot_);
+
+  wal_incremental_seq = storage_->GetDB()->GetLatestSequenceNumber();
+  if (wal_incremental_seq > wal_begin_seq_) {
+    auto s = migrateIncrementalDataByRawKV(wal_incremental_seq, &batch_sender);
+    if (!s.IsOK()) {
+      return {Status::NotOK, fmt::format("migrate last incremental data failed, {}", s.Msg())};
+    }
+    LOG(INFO) << fmt::format("[migrate] Migrated last incremental data after set forbidden slot, seq from {} to {}",
+                             wal_begin_seq_, wal_incremental_seq);
+  }
+
+  auto elapsed = util::GetTimeStampMS() - start_ts;
+  LOG(INFO) << fmt::format(
+      "[migrate] Succeed to migrate incremental data, slot: {}, elapsed: {} ms, "
+      "sent: {} bytes, rate: {:.2f} kb/s, batches: {}, entries: {}",
+      migrating_slot_.load(), elapsed, batch_sender.GetSentBytes(), batch_sender.GetRate(start_ts),
+      batch_sender.GetSentBatchesNum(), batch_sender.GetEntriesNum());
+
+  return Status::OK();
+}
+
+bool SlotMigrator::catchUpIncrementalWAL() {
+  uint64_t gap = storage_->GetDB()->GetLatestSequenceNumber() - wal_begin_seq_;
+  if (gap <= seq_gap_limit_) {
+    LOG(INFO) << fmt::format("[migrate] Incremental data sequence gap: {}, less than limit: {}, set forbidden slot: {}",
+                             gap, seq_gap_limit_, migrating_slot_.load());
+    return true;
+  }
+  return false;
+}
+
+Status SlotMigrator::migrateIncrementalDataByRawKV(uint64_t end_seq, BatchSender *batch_sender) {
+  engine::WALIterator wal_iter(storage_, migrating_slot_);
+  uint64_t start_seq = wal_begin_seq_ + 1;
+  for (wal_iter.Seek(start_seq); wal_iter.Valid(); wal_iter.Next()) {
+    if (wal_iter.NextSequenceNumber() > end_seq + 1) {
+      break;
+    }
+    auto item = wal_iter.Item();
+    switch (item.type) {
+      case engine::WALItem::Type::kTypeLogData: {
+        GET_OR_RET(batch_sender->PutLogData(item.key));
+        break;
+      }
+      case engine::WALItem::Type::kTypePut: {
+        GET_OR_RET(batch_sender->Put(storage_->GetCFHandle(static_cast<ColumnFamilyID>(item.column_family_id)),
+                                     item.key, item.value));
+        break;
+      }
+      case engine::WALItem::Type::kTypeDelete: {
+        GET_OR_RET(
+            batch_sender->Delete(storage_->GetCFHandle(static_cast<ColumnFamilyID>(item.column_family_id)), item.key));
+        break;
+      }
+      default:
```

Comment:
I don't have a good way to handle delete range because it may overwrite multiple slots. Do you have any good ideas? 
I don't see that the redis-command migration handles delete range either.

Reply:
Sound good :)